File: backend/expression_eval.py
"""Safe expression evaluator for computed interview fields.

Supports:
    - Arithmetic: +, -, *, /
    - Numeric literals: 42, 3.14
    - Field references: field_id (top-level), group.child (repeat-group member)
    - Aggregate functions over repeat groups:
        sum(group.field)            — sum of a single field
        sum(group.a * group.b)      — sum of per-row products
        count(group)                — number of items
        avg(group.field)            — average
        min(group.field)            — minimum
        max(group.field)            — maximum
    - Parentheses for grouping
    - round(expr, digits)           — round to N decimal places

All evaluation is done via a recursive-descent parser — no eval().
"""

from __future__ import annotations

import re
from typing import Any

# ---------------------------------------------------------------------------
# Tokeniser
# ---------------------------------------------------------------------------

_TOKEN_RE = re.compile(
    r"""
    (\d+(?:\.\d+)?)      |  # number literal
    ([a-zA-Z_]\w*(?:\.[a-zA-Z_]\w*)*)  |  # identifier / dotted ref
    ([+\-*/(),])            # operator / punctuation
    """,
    re.VERBOSE,
)

_TOK_NUM = "NUM"
_TOK_ID = "ID"
_TOK_OP = "OP"

_AGGREGATES = {"sum", "count", "avg", "min", "max"}
_FUNCTIONS = _AGGREGATES | {"round"}


def _tokenise(expr: str) -> list[tuple[str, str]]:
    tokens: list[tuple[str, str]] = []
    for m in _TOKEN_RE.finditer(expr):
        if m.group(1):
            tokens.append((_TOK_NUM, m.group(1)))
        elif m.group(2):
            tokens.append((_TOK_ID, m.group(2)))
        elif m.group(3):
            tokens.append((_TOK_OP, m.group(3)))
    return tokens


# ---------------------------------------------------------------------------
# Parser  (recursive-descent, produces an AST)
# ---------------------------------------------------------------------------

class _Parser:
    """Recursive-descent parser producing a simple AST."""

    def __init__(self, tokens: list[tuple[str, str]]) -> None:
        self.tokens = tokens
        self.pos = 0

    def _peek(self) -> tuple[str, str] | None:
        if self.pos < len(self.tokens):
            return self.tokens[self.pos]
        return None

    def _advance(self) -> tuple[str, str]:
        tok = self.tokens[self.pos]
        self.pos += 1
        return tok

    def _expect(self, kind: str, value: str | None = None) -> tuple[str, str]:
        tok = self._peek()
        if tok is None:
            raise ExpressionError(f"Unexpected end of expression, expected {value or kind}")
        if tok[0] != kind or (value is not None and tok[1] != value):
            raise ExpressionError(f"Expected '{value or kind}', got '{tok[1]}'")
        return self._advance()

    # expr  -> term (('+' | '-') term)*
    def parse_expr(self) -> dict:
        node = self._parse_term()
        while True:
            tok = self._peek()
            if tok and tok[0] == _TOK_OP and tok[1] in ("+", "-"):
                op = self._advance()[1]
                right = self._parse_term()
                node = {"op": op, "left": node, "right": right}
            else:
                break
        return node

    # term  -> unary (('*' | '/') unary)*
    def _parse_term(self) -> dict:
        node = self._parse_unary()
        while True:
            tok = self._peek()
            if tok and tok[0] == _TOK_OP and tok[1] in ("*", "/"):
                op = self._advance()[1]
                right = self._parse_unary()
                node = {"op": op, "left": node, "right": right}
            else:
                break
        return node

    # unary -> '-' unary | atom
    def _parse_unary(self) -> dict:
        tok = self._peek()
        if tok and tok[0] == _TOK_OP and tok[1] == "-":
            self._advance()
            operand = self._parse_unary()
            return {"op": "neg", "operand": operand}
        return self._parse_atom()

    # atom  -> NUMBER | function_call | identifier | '(' expr ')'
    def _parse_atom(self) -> dict:
        tok = self._peek()
        if tok is None:
            raise ExpressionError("Unexpected end of expression")

        if tok[0] == _TOK_NUM:
            self._advance()
            return {"lit": float(tok[1])}

        if tok[0] == _TOK_ID:
            name = tok[1]
            self._advance()
            # Check if it's a function call
            nxt = self._peek()
            if nxt and nxt[0] == _TOK_OP and nxt[1] == "(":
                if name not in _FUNCTIONS:
                    raise ExpressionError(f"Unknown function '{name}'")
                self._advance()  # consume '('
                if name == "count":
                    # count(group) — argument is a single identifier
                    arg_tok = self._expect(_TOK_ID)
                    self._expect(_TOK_OP, ")")
                    return {"func": "count", "group": arg_tok[1]}
                elif name == "round":
                    # round(expr, digits)
                    arg = self.parse_expr()
                    self._expect(_TOK_OP, ",")
                    digits_tok = self._expect(_TOK_NUM)
                    self._expect(_TOK_OP, ")")
                    return {"func": "round", "arg": arg, "digits": int(float(digits_tok[1]))}
                else:
                    # sum/avg/min/max — argument is an expression (parsed in aggregate context)
                    arg = self.parse_expr()
                    self._expect(_TOK_OP, ")")
                    return {"func": name, "arg": arg}
            # Plain identifier (field reference)
            return {"ref": name}

        if tok[0] == _TOK_OP and tok[1] == "(":
            self._advance()
            node = self.parse_expr()
            self._expect(_TOK_OP, ")")
            return node

        raise ExpressionError(f"Unexpected token '{tok[1]}'")


# ---------------------------------------------------------------------------
# Evaluator
# ---------------------------------------------------------------------------

class ExpressionError(Exception):
    """Raised for malformed or unevaluable expressions."""


def _to_number(v: Any) -> float:
    if v is None or v == "":
        return 0.0
    try:
        return float(v)
    except (ValueError, TypeError):
        return 0.0


def _resolve_ref(ref: str, data: dict) -> Any:
    """Resolve a dotted reference against the submission data.

    - ``total_price``       → ``data["total_price"]``
    - ``actions.unit_price``→ raises — only valid inside an aggregate
    """
    if "." in ref:
        raise ExpressionError(
            f"Dotted reference '{ref}' can only be used inside an aggregate function "
            f"like sum(), avg(), etc."
        )
    return data.get(ref)


def _collect_groups(node: dict) -> set[str]:
    """Walk an AST node and return the set of repeat-group names referenced via dot notation."""
    groups: set[str] = set()
    if "ref" in node:
        if "." in node["ref"]:
            groups.add(node["ref"].split(".")[0])
    elif "left" in node:
        groups |= _collect_groups(node["left"])
        groups |= _collect_groups(node["right"])
    elif "operand" in node:
        groups |= _collect_groups(node["operand"])
    elif "arg" in node and isinstance(node["arg"], dict):
        groups |= _collect_groups(node["arg"])
    return groups


def _eval_row(node: dict, row: dict, data: dict) -> float:
    """Evaluate an AST node in the context of a single repeat-group row."""
    if "lit" in node:
        return node["lit"]

    if "ref" in node:
        ref = node["ref"]
        if "." in ref:
            _group, field = ref.split(".", 1)
            return _to_number(row.get(field))
        return _to_number(data.get(ref))

    if "op" in node:
        op = node["op"]
        if op == "neg":
            return -_eval_row(node["operand"], row, data)
        left = _eval_row(node["left"], row, data)
        right = _eval_row(node["right"], row, data)
        if op == "+":
            return left + right
        if op == "-":
            return left - right
        if op == "*":
            return left * right
        if op == "/":
            return left / right if right != 0 else 0.0

    if "func" in node:
        return _eval_node(node, data)

    raise ExpressionError(f"Cannot evaluate node: {node}")


def _eval_node(node: dict, data: dict) -> float:
    """Evaluate an AST node against the full submission data."""
    if "lit" in node:
        return node["lit"]

    if "ref" in node:
        return _to_number(_resolve_ref(node["ref"], data))

    if "op" in node:
        op = node["op"]
        if op == "neg":
            return -_eval_node(node["operand"], data)
        left = _eval_node(node["left"], data)
        right = _eval_node(node["right"], data)
        if op == "+":
            return left + right
        if op == "-":
            return left - right
        if op == "*":
            return left * right
        if op == "/":
            return left / right if right != 0 else 0.0

    if "func" in node:
        func = node["func"]

        if func == "count":
            group_name = node["group"]
            items = data.get(group_name, [])
            return float(len(items) if isinstance(items, list) else 0)

        if func == "round":
            val = _eval_node(node["arg"], data)
            return round(val, node["digits"])

        # Aggregate: sum, avg, min, max
        arg = node["arg"]
        groups = _collect_groups(arg)
        if not groups:
            raise ExpressionError(
                f"{func}() requires a repeat-group reference (e.g. {func}(group.field))"
            )
        if len(groups) > 1:
            raise ExpressionError(
                f"{func}() references multiple groups: {groups}. Only one is allowed."
            )
        group_name = groups.pop()
        items = data.get(group_name, [])
        if not isinstance(items, list):
            items = []

        if len(items) == 0:
            return 0.0

        values = [_eval_row(arg, row, data) for row in items]

        if func == "sum":
            return sum(values)
        if func == "avg":
            return sum(values) / len(values)
        if func == "min":
            return min(values)
        if func == "max":
            return max(values)

    raise ExpressionError(f"Cannot evaluate node: {node}")


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def parse_expression(expr: str) -> dict:
    """Parse an expression string into an AST."""
    tokens = _tokenise(expr)
    if not tokens:
        raise ExpressionError("Empty expression")
    parser = _Parser(tokens)
    ast = parser.parse_expr()
    if parser.pos < len(parser.tokens):
        raise ExpressionError(
            f"Unexpected token '{parser.tokens[parser.pos][1]}' after expression"
        )
    return ast


def evaluate_expression(expr: str, data: dict) -> float:
    """Parse and evaluate an expression against submission data.

    Returns the computed numeric result.
    """
    ast = parse_expression(expr)
    return _eval_node(ast, data)


def evaluate_computed_fields(components: list, data: dict) -> dict:
    """Evaluate all computed (expression) fields and return an updated data dict.

    Iterates components (including nested ones) looking for number fields
    with an ``expression`` property.  Computed values are injected into a
    *copy* of *data* so they are available for docx rendering and
    subsequent expressions.
    """
    result = dict(data)
    _eval_components(components, result)
    return result


def _eval_components(components: list, data: dict) -> None:
    """Recursively evaluate expression fields in component list."""
    for comp in components:
        ctype = comp.get("type", "")
        if ctype in ("dialog", "repeat"):
            nested = comp.get("components", [])
            if ctype == "dialog":
                _eval_components(nested, data)
            # For repeat groups, evaluate expressions inside each row
            if ctype == "repeat":
                group_id = comp.get("id", "")
                items = data.get(group_id, [])
                if isinstance(items, list):
                    for row in items:
                        _eval_components(nested, {**data, **row})
                        # Write computed values back into the row
                        for child in nested:
                            if child.get("type") == "number" and child.get("expression"):
                                row[child["id"]] = evaluate_expression(
                                    child["expression"], {**data, **row}
                                )
        elif ctype == "number" and comp.get("expression"):
            field_id = comp["id"]
            try:
                data[field_id] = evaluate_expression(comp["expression"], data)
            except ExpressionError:
                pass  # leave as-is if expression can't be evaluated

File: frontend/src/expressionEval.js
/**
 * Safe expression evaluator for computed interview fields.
 *
 * Supports:
 *   - Arithmetic: +, -, *, /
 *   - Numeric literals: 42, 3.14
 *   - Field references: fieldId (top-level), group.child (repeat-group)
 *   - Aggregate functions: sum(), count(), avg(), min(), max()
 *   - round(expr, digits)
 *   - Parentheses for grouping
 *
 * Mirrors the backend expression_eval.py — no eval().
 */

const TOKEN_RE = /(\d+(?:\.\d+)?)|([a-zA-Z_]\w*(?:\.[a-zA-Z_]\w*)*)|([+\-*/(),])/g

const FUNCTIONS = new Set(['sum', 'count', 'avg', 'min', 'max', 'round'])
const AGGREGATES = new Set(['sum', 'count', 'avg', 'min', 'max'])

function tokenise(expr) {
  const tokens = []
  let m
  TOKEN_RE.lastIndex = 0
  while ((m = TOKEN_RE.exec(expr)) !== null) {
    if (m[1] !== undefined) tokens.push({ t: 'NUM', v: m[1] })
    else if (m[2] !== undefined) tokens.push({ t: 'ID', v: m[2] })
    else if (m[3] !== undefined) tokens.push({ t: 'OP', v: m[3] })
  }
  return tokens
}

class Parser {
  constructor(tokens) {
    this.tokens = tokens
    this.pos = 0
  }

  peek() { return this.pos < this.tokens.length ? this.tokens[this.pos] : null }
  advance() { return this.tokens[this.pos++] }
  expect(t, v) {
    const tok = this.peek()
    if (!tok || tok.t !== t || (v !== undefined && tok.v !== v))
      throw new Error(`Expected '${v || t}', got '${tok ? tok.v : 'EOF'}'`)
    return this.advance()
  }

  parseExpr() {
    let node = this.parseTerm()
    while (true) {
      const tok = this.peek()
      if (tok && tok.t === 'OP' && (tok.v === '+' || tok.v === '-')) {
        const op = this.advance().v
        node = { op, left: node, right: this.parseTerm() }
      } else break
    }
    return node
  }

  parseTerm() {
    let node = this.parseUnary()
    while (true) {
      const tok = this.peek()
      if (tok && tok.t === 'OP' && (tok.v === '*' || tok.v === '/')) {
        const op = this.advance().v
        node = { op, left: node, right: this.parseUnary() }
      } else break
    }
    return node
  }

  parseUnary() {
    const tok = this.peek()
    if (tok && tok.t === 'OP' && tok.v === '-') {
      this.advance()
      return { op: 'neg', operand: this.parseUnary() }
    }
    return this.parseAtom()
  }

  parseAtom() {
    const tok = this.peek()
    if (!tok) throw new Error('Unexpected end of expression')

    if (tok.t === 'NUM') {
      this.advance()
      return { lit: parseFloat(tok.v) }
    }

    if (tok.t === 'ID') {
      const name = tok.v
      this.advance()
      const nxt = this.peek()
      if (nxt && nxt.t === 'OP' && nxt.v === '(') {
        if (!FUNCTIONS.has(name)) throw new Error(`Unknown function '${name}'`)
        this.advance() // consume '('
        if (name === 'count') {
          const arg = this.expect('ID')
          this.expect('OP', ')')
          return { func: 'count', group: arg.v }
        }
        if (name === 'round') {
          const arg = this.parseExpr()
          this.expect('OP', ',')
          const digits = this.expect('NUM')
          this.expect('OP', ')')
          return { func: 'round', arg, digits: parseInt(digits.v) }
        }
        // sum/avg/min/max
        const arg = this.parseExpr()
        this.expect('OP', ')')
        return { func: name, arg }
      }
      return { ref: name }
    }

    if (tok.t === 'OP' && tok.v === '(') {
      this.advance()
      const node = this.parseExpr()
      this.expect('OP', ')')
      return node
    }

    throw new Error(`Unexpected token '${tok.v}'`)
  }
}

function toNum(v) {
  if (v == null || v === '') return 0
  const n = parseFloat(v)
  return isNaN(n) ? 0 : n
}

function collectGroups(node) {
  const groups = new Set()
  if (node.ref && node.ref.includes('.')) groups.add(node.ref.split('.')[0])
  if (node.left) { collectGroups(node.left).forEach(g => groups.add(g)); collectGroups(node.right).forEach(g => groups.add(g)) }
  if (node.operand) collectGroups(node.operand).forEach(g => groups.add(g))
  if (node.arg && typeof node.arg === 'object') collectGroups(node.arg).forEach(g => groups.add(g))
  return groups
}

function evalRow(node, row, data) {
  if ('lit' in node) return node.lit
  if ('ref' in node) {
    if (node.ref.includes('.')) {
      const field = node.ref.split('.').slice(1).join('.')
      return toNum(row[field])
    }
    return toNum(data[node.ref])
  }
  if ('op' in node) {
    if (node.op === 'neg') return -evalRow(node.operand, row, data)
    const l = evalRow(node.left, row, data)
    const r = evalRow(node.right, row, data)
    if (node.op === '+') return l + r
    if (node.op === '-') return l - r
    if (node.op === '*') return l * r
    if (node.op === '/') return r !== 0 ? l / r : 0
  }
  if ('func' in node) return evalNode(node, data)
  return 0
}

function evalNode(node, data) {
  if ('lit' in node) return node.lit
  if ('ref' in node) {
    if (node.ref.includes('.')) return 0 // dotted ref outside aggregate
    return toNum(data[node.ref])
  }
  if ('op' in node) {
    if (node.op === 'neg') return -evalNode(node.operand, data)
    const l = evalNode(node.left, data)
    const r = evalNode(node.right, data)
    if (node.op === '+') return l + r
    if (node.op === '-') return l - r
    if (node.op === '*') return l * r
    if (node.op === '/') return r !== 0 ? l / r : 0
  }
  if ('func' in node) {
    const { func } = node
    if (func === 'count') {
      const items = data[node.group]
      return Array.isArray(items) ? items.length : 0
    }
    if (func === 'round') {
      const val = evalNode(node.arg, data)
      const factor = Math.pow(10, node.digits)
      return Math.round(val * factor) / factor
    }
    // sum/avg/min/max
    const groups = collectGroups(node.arg)
    if (groups.size === 0) return 0
    const groupName = [...groups][0]
    const items = data[groupName]
    if (!Array.isArray(items) || items.length === 0) return 0
    const values = items.map(row => evalRow(node.arg, row, data))
    if (func === 'sum') return values.reduce((a, b) => a + b, 0)
    if (func === 'avg') return values.reduce((a, b) => a + b, 0) / values.length
    if (func === 'min') return Math.min(...values)
    if (func === 'max') return Math.max(...values)
  }
  return 0
}

export function evaluateExpression(expr, data) {
  try {
    const tokens = tokenise(expr)
    if (tokens.length === 0) return null
    const parser = new Parser(tokens)
    const ast = parser.parseExpr()
    return evalNode(ast, data)
  } catch {
    return null
  }
}

/**
 * Evaluate all computed fields in a component list and return updated data.
 */
export function evaluateComputedFields(components, data) {
  const result = { ...data }
  evalComponents(components, result)
  return result
}

function evalComponents(components, data) {
  for (const comp of components) {
    if (comp.type === 'dialog') {
      evalComponents(comp.components || [], data)
    } else if (comp.type === 'repeat') {
      const items = data[comp.id]
      if (Array.isArray(items)) {
        for (const row of items) {
          for (const child of (comp.components || [])) {
            if (child.type === 'number' && child.expression) {
              const val = evaluateExpression(child.expression, { ...data, ...row })
              if (val !== null) row[child.id] = val
            }
          }
        }
      }
    } else if (comp.type === 'number' && comp.expression) {
      const val = evaluateExpression(comp.expression, data)
      if (val !== null) data[comp.id] = val
    }
  }
}
